File: scripts/enforce-licenses.js
#!/usr/bin/env node
/**
 * enforce-licenses.js <licenses-node.json> <allowlist.json>
 *
 * Reads license-checker output (JSON), checks each package license against
 * the allow/deny list. Exits 1 on any denied license.
 */
const fs = require('fs');
const [,, licensesPath, allowlistPath] = process.argv;
if (!licensesPath || !allowlistPath) {
  console.error('usage: enforce-licenses.js <licenses.json> <allowlist.json>');
  process.exit(2);
}
const licenses = JSON.parse(fs.readFileSync(licensesPath, 'utf8'));
const { allowed, denied } = JSON.parse(fs.readFileSync(allowlistPath, 'utf8'));

const isDenied = (l) => denied.some(d => new RegExp('^' + d.replace(/\*/g,'.*') + '$', 'i').test(l));
const isAllowed = (l) => allowed.some(a => a.toLowerCase() === l.toLowerCase());

let fail = 0;
for (const [pkg, meta] of Object.entries(licenses)) {
  const raw = meta.licenses || 'UNKNOWN';
  const list = Array.isArray(raw) ? raw : String(raw).split(/\s+OR\s+|,\s*/i);
  const passes = list.some(isAllowed);
  const blocked = list.some(isDenied);
  if (blocked) {
    console.error(`::error::license DENIED: ${pkg} (${raw})`);
    fail++;
  } else if (!passes) {
    console.error(`::error::license UNAPPROVED: ${pkg} (${raw}) — request Legal review or add to allowlist`);
    fail++;
  }
}
if (fail) {
  console.error(`\nlicense enforcement failed: ${fail} package(s)`);
  process.exit(1);
}
console.log('license enforcement: OK');

File: scripts/enforce_licenses.py
#!/usr/bin/env python3
"""enforce_licenses.py <licenses.json> <allowlist.json>

Reads pip-licenses JSON output and enforces the org allowlist.
"""
import json, re, sys

def norm(x: str) -> str:
    return re.sub(r"\s+", "", (x or "")).lower()

def main(lic_path: str, allow_path: str) -> int:
    licenses = json.load(open(lic_path))
    allow = json.load(open(allow_path))
    allowed = {norm(a) for a in allow["allowed"]}
    denied = [re.compile("^" + a.replace("*", ".*") + "$", re.I) for a in allow["denied"]]
    fail = 0
    for p in licenses:
        name = p.get("Name", "<unknown>")
        lic = p.get("License", "UNKNOWN")
        parts = re.split(r"\s+OR\s+|,\s*|\s+AND\s+", lic)
        blocked = any(any(pat.match(x.strip()) for pat in denied) for x in parts)
        ok = any(norm(x) in allowed for x in parts)
        if blocked:
            print(f"::error::license DENIED: {name} ({lic})"); fail += 1
        elif not ok:
            print(f"::error::license UNAPPROVED: {name} ({lic}) — request Legal review or add to allowlist"); fail += 1
    if fail:
        print(f"license enforcement failed: {fail} package(s)"); return 1
    print("license enforcement: OK")
    return 0

if __name__ == "__main__":
    sys.exit(main(sys.argv[1], sys.argv[2]))
